// @gnomad/ui/lib/cjs/Checkbox.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Checkbox = void 0;
var _propTypes = _interopRequireDefault(require("prop-types"));
var _react = _interopRequireDefault(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _excluded = ["checked", "disabled", "id", "label", "onChange"];
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
var Label = _styledComponents.default.label.withConfig({
  displayName: "Checkbox__Label",
  componentId: "sc-hrpa6b-0"
})(["user-select:none;"]);
var CheckboxInput = _styledComponents.default.input.withConfig({
  displayName: "Checkbox__CheckboxInput",
  componentId: "sc-hrpa6b-1"
})(["margin-right:0.5em;"]);
var Checkbox = function Checkbox(_ref) {
  var checked = _ref.checked,
    disabled = _ref.disabled,
    id = _ref.id,
    label = _ref.label,
    _onChange = _ref.onChange,
    rest = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/_react.default.createElement(Label, _extends({}, rest, {
    htmlFor: id
  }), /*#__PURE__*/_react.default.createElement(CheckboxInput, {
    checked: checked,
    disabled: disabled,
    id: id,
    onChange: function onChange(e) {
      _onChange(e.target.checked);
    },
    type: "checkbox"
  }), label);
};
exports.Checkbox = Checkbox;
Checkbox.propTypes = {
  checked: _propTypes.default.bool.isRequired,
  disabled: _propTypes.default.bool,
  id: _propTypes.default.string.isRequired,
  label: _propTypes.default.string.isRequired,
  onChange: _propTypes.default.func.isRequired
};
Checkbox.defaultProps = {
  disabled: false
};
