// @gnomad/track-genes/lib/cjs/GenesTrack.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GenesTrack = void 0;
var _propTypes = _interopRequireDefault(require("prop-types"));
var _react = _interopRequireDefault(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _regionViewer = require("@gnomad/region-viewer");
var _GenesPlot = require("./GenesPlot");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var TitlePanel = _styledComponents.default.div.withConfig({
  displayName: "GenesTrack__TitlePanel",
  componentId: "sc-189pu5x-0"
})(["display:flex;flex-direction:column;justify-content:flex-start;"]);
var GenesTrack = function GenesTrack(_ref) {
  var genes = _ref.genes,
    includeNonCodingGenes = _ref.includeNonCodingGenes,
    renderGeneLabel = _ref.renderGeneLabel,
    title = _ref.title;
  return /*#__PURE__*/_react.default.createElement(_regionViewer.Track, {
    renderLeftPanel: function renderLeftPanel() {
      return /*#__PURE__*/_react.default.createElement(TitlePanel, null, title);
    }
  }, function (_ref2) {
    var scalePosition = _ref2.scalePosition,
      width = _ref2.width;
    return /*#__PURE__*/_react.default.createElement(_GenesPlot.GenesPlot, {
      genes: genes,
      includeNonCodingGenes: includeNonCodingGenes,
      renderGeneLabel: renderGeneLabel,
      scalePosition: scalePosition,
      width: width
    });
  });
};
exports.GenesTrack = GenesTrack;
GenesTrack.propTypes = {
  genes: _propTypes.default.arrayOf(_propTypes.default.shape({
    gene_id: _propTypes.default.string.isRequired,
    start: _propTypes.default.number.isRequired,
    stop: _propTypes.default.number.isRequired,
    exons: _propTypes.default.arrayOf(_propTypes.default.shape({
      feature_type: _propTypes.default.oneOf(['CDS', 'exon', 'UTR']).isRequired,
      start: _propTypes.default.number.isRequired,
      stop: _propTypes.default.number.isRequired
    })).isRequired
  })).isRequired,
  includeNonCodingGenes: _propTypes.default.bool,
  renderGeneLabel: _propTypes.default.func,
  title: _propTypes.default.string
};
GenesTrack.defaultProps = {
  includeNonCodingGenes: false,
  renderGeneLabel: undefined,
  title: 'Genes'
};
