// @gnomad/region-viewer/lib/cjs/Track.js
"use strict";

function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Track = void 0;
var _propTypes = _interopRequireDefault(require("prop-types"));
var _react = _interopRequireDefault(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _RegionViewer = require("./RegionViewer");
var _excluded = ["children", "renderLeftPanel", "renderRightPanel", "renderTopPanel"];
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
var OuterWrapper = _styledComponents.default.div.withConfig({
  displayName: "Track__OuterWrapper",
  componentId: "sc-1sdyh2h-0"
})(["display:flex;flex-direction:column;"]);
var InnerWrapper = _styledComponents.default.div.withConfig({
  displayName: "Track__InnerWrapper",
  componentId: "sc-1sdyh2h-1"
})(["display:flex;flex-direction:row;align-items:stretch;"]);
var TopPanel = _styledComponents.default.div.withConfig({
  displayName: "Track__TopPanel",
  componentId: "sc-1sdyh2h-2"
})(["display:flex;"]);
var SidePanel = _styledComponents.default.div.withConfig({
  displayName: "Track__SidePanel",
  componentId: "sc-1sdyh2h-3"
})(["display:flex;flex-shrink:0;flex-direction:column;"]);
var CenterPanel = _styledComponents.default.div.withConfig({
  displayName: "Track__CenterPanel",
  componentId: "sc-1sdyh2h-4"
})(["display:flex;flex-shrink:0;flex-direction:column;"]);
var TitlePanel = _styledComponents.default.div.withConfig({
  displayName: "Track__TitlePanel",
  componentId: "sc-1sdyh2h-5"
})(["display:flex;flex-direction:column;justify-content:center;height:100%;"]);

// eslint-disable-next-line react/prop-types
var defaultRenderLeftPanel = function defaultRenderLeftPanel(_ref) {
  var _ref$title = _ref.title,
    title = _ref$title === void 0 ? '' : _ref$title;
  return /*#__PURE__*/_react.default.createElement(TitlePanel, null, title.split('\n').map(function (s) {
    return /*#__PURE__*/_react.default.createElement("span", {
      key: s
    }, s);
  }));
};
var Track = function Track(_ref2) {
  var children = _ref2.children,
    renderLeftPanel = _ref2.renderLeftPanel,
    renderRightPanel = _ref2.renderRightPanel,
    renderTopPanel = _ref2.renderTopPanel,
    rest = _objectWithoutProperties(_ref2, _excluded);
  return /*#__PURE__*/_react.default.createElement(_RegionViewer.RegionViewerContext.Consumer, null, function (_ref3) {
    var centerPanelWidth = _ref3.centerPanelWidth,
      isPositionDefined = _ref3.isPositionDefined,
      leftPanelWidth = _ref3.leftPanelWidth,
      regions = _ref3.regions,
      rightPanelWidth = _ref3.rightPanelWidth,
      scalePosition = _ref3.scalePosition;
    return /*#__PURE__*/_react.default.createElement(OuterWrapper, null, renderTopPanel && /*#__PURE__*/_react.default.createElement(TopPanel, {
      style: {
        width: centerPanelWidth,
        marginLeft: leftPanelWidth,
        marginRight: rightPanelWidth
      }
    }, renderTopPanel(_objectSpread(_objectSpread({}, rest), {}, {
      width: centerPanelWidth
    }))), /*#__PURE__*/_react.default.createElement(InnerWrapper, null, /*#__PURE__*/_react.default.createElement(SidePanel, {
      style: {
        width: leftPanelWidth
      }
    }, renderLeftPanel && renderLeftPanel(_objectSpread(_objectSpread({}, rest), {}, {
      width: leftPanelWidth
    }))), /*#__PURE__*/_react.default.createElement(CenterPanel, {
      style: {
        width: centerPanelWidth
      }
    }, children(_objectSpread(_objectSpread({}, rest), {}, {
      isPositionDefined: isPositionDefined,
      leftPanelWidth: leftPanelWidth,
      regions: regions,
      rightPanelWidth: rightPanelWidth,
      scalePosition: scalePosition,
      width: centerPanelWidth
    }))), renderRightPanel && /*#__PURE__*/_react.default.createElement(SidePanel, {
      style: {
        width: rightPanelWidth
      }
    }, renderRightPanel(_objectSpread(_objectSpread({}, rest), {}, {
      width: rightPanelWidth
    })))));
  });
};
exports.Track = Track;
Track.propTypes = {
  children: _propTypes.default.func.isRequired,
  renderLeftPanel: _propTypes.default.func,
  renderRightPanel: _propTypes.default.func,
  renderTopPanel: _propTypes.default.func
};
Track.defaultProps = {
  renderLeftPanel: defaultRenderLeftPanel,
  renderRightPanel: null,
  renderTopPanel: null
};
