---
question: 'What are NUMTs and why are they problematic?'
---

Nuclear sequences of mitochondrial origin (NUMTs) are derived from pieces of mtDNA that have integrated into the nuclear genome over the course of human evolution. Many ancient NUMTs are part of the reference human genome assembly; however, hundreds of polymorphic NUMTs exist that are present only in some individuals. Reads derived from NUMTs (particularly polymorphic NUMTs) often mis-align to the mtDNA and conversely, reads genuinely arising from the mtDNA genome can mis-align to the NUMTs in the reference genome. NUMTs present problems for calling variants in the mtDNA because reads that mis-align to the mtDNA will often generate false positive calls. For example, in a sample with 30x autosomal coverage and 1500x mtDNA coverage, a homozygous NUMT that mis-aligns to the mtDNA can result in false positive calls at 2% apparent heteroplasmy (30/1530 reads). Other bases in the same NUMT can cause a truly homoplasmic mtDNA allele to be falsely called at 98% heteroplasmy (with the 2% REF alleles deriving from the NUMT mis-alignment). This is one reason why we call all variants 95-100% heteroplasmy as "homoplasmic or near-homoplasmic" -- since the REF alleles could be derived from NUMTs or from sequencing errors. NUMTs are a larger problem for samples with few mtDNA molecules/cell, which is why we exclude samples with < 50 mtDNA copies/cell.
